src/libostree/ostree-repo.c

```diff
@@ -435,16 +456,20 @@ pop_repo_lock (OstreeRepo  *self,
     {
       /* Lock stack will be empty, unlock */
       g_debug ("Unlocking repo");
-      if (!do_repo_unlock (lock->fd, flags))
+      if (!do_repo_unlock (self->lock.fd, flags))
         return glnx_throw_errno_prefix (error, "Unlocking repo failed");
     }
 
-  g_queue_pop_head (&lock->stack);
+  if (state_to_drop == LOCK_EX)
+    self->lock.exclusive--;
+  else
+    self->lock.shared--;
+  g_ptr_array_remove_index (stack, stack->len - 1);
```

Comment:
Hmm wait, now that we have the two counters, do we actually need to store the stack?  Since the valid state transitions can only be:

shared (++) -> exclusive (++)

i.e. we can never have any shared *after* exclusive, so why maintain the stack at all versus just the two counters?

Reply:
I briefly came to the same conclusion, but I think it's necessary to know what state to drop when there are multiple threads. In fact, what's here has a bug and I'm going to beef up the multithreaded test to check this.

Here's the scenario if I can get this ascii timeline to work out:
```
Thread 1	Thread 2	Flock state
========	========	===========
push exclusive			exclusive
		push shared	exclusive
pop				shared
```
With the current implementation, the flock would remain exclusive after the pop because the counters would say `exclusive == 2` and `shared == 0` before. In order to correctly return from exclusive to shared when thread 1 pops, you need to correctly track how many exclusive and shared locks are held instead of just incrementing `exclusive` every time once the lock is exclusive. Then it would be `exclusive == 1` and `shared == 1`, but you still wouldn't know which state to drop unless it was stored with the thread that is doing the dropping.

Reply:
I just pushed 3f78672 which tests this scenario and will now fail.

Reply:
That failed as expected:
```
[2021-04-27T21:19:26.049Z] ERROR: tests/test-repo - Bail out! OSTree:ERROR:tests/test-repo.c:459:test_repo_lock_multi_thread: assertion failed (error == NULL): Locking repo shared failed: Resource temporarily unavailable (g-io-error-quark, 27)
```

Reply:
And now with c7de18a `tests/test-repo` passes. Unless you still think that's incorrect, I'm going to clean up the series.

Reply:
> In order to correctly return from exclusive to shared when thread 1 pops, you need to correctly track how many exclusive and shared locks are held instead of just incrementing exclusive every time once the lock is exclusive. Then it would be exclusive == 1 and shared == 1, but you still wouldn't know which state to drop unless it was stored with the thread that is doing the dropping.

I need to think about this a bit more.  So what we're trying to support is a case where higher level code grabs an exclusive lock, but then some portion of that operation wants a shared lock, right?  And the idea here is that everything that has access to the same `OstreeRepo` instance is part of the same process.  

That part makes sense to me.

But your state diagram has *threads* - it should be separate processes and `OstreeRepo*` instances right?

Reply:
It would be great to only think about processes and just have a mutex that guards the lock state. Unfortunately, once you're running threads the order that the lock is taken and released is no longer linear. 2 worker threads aren't going to coordinate to release their locks in the same order they were acquired. And since the lock has multiple states, you need to know what each particular thread is trying to release. If it was just a single lock/unlock state like a mutex, then you could just have a single counter and when it gets to 0 you unlock.

I honestly wouldn't have given multithreaded code using the same `OstreeRepo` much thought except that the pull code already spawns a bunch of threads. So even in ostree you could trigger this behavior just by taking an exclusive lock somewhere and then running some pull operation _if_ a lock was taken by any of the pull threads.

What actually tipped me off to this was that I wanted to make `ostree_repo_prune_object` take an exclusive lock. In the vast majority of cases that works fine - `ostree_repo_prune` takes an exclusive lock and the lock increments each time `ostree_repo_prune_object` is called. Except that when you try to write a commit object it also tries to prune a potentially existing old tombstone commit with `ostree_repo_prune_object`. That immediately caused pull to deadlock since the pull thread's lock (with what's in ostree right now) is disconnected from the main thread. OK, let's make them coordinate by keeping the lock in the shared `OstreeRepo`. Oh wait, that's not so simple to track correctly.

An alternative I just thought of that would require less state tracking in ostree would be to have the caller specify the state to drop in the `pop()` call. Then you could have 2 counters in the `OstreeRepo` and just decrement what the caller said without the per-thread stack keeping track of what the thread's last push call was. I don't really know what that API would look like.

Reply:
But all that locking is internal to ostree today; we can just add an `_ostree_repo_prune_object_unlocked()` that assumes the caller has locked, and use that in the pull code right?

Reply:
I don't know what happened to the comment I made the other day.

Yes, in that case you could add some unlocked variant that's only available internally. That case actually violates the semantics since the transaction only has a shared lock and the expectation is that you have an exclusive lock when deleting. Regardless, I think in general once you allow the locking to be made externally you will sooner or later run into a case where someone is taking a lock externally in a way that the ostree internals haven't arranged for. To me the only real fix is to make the locking completely safe to use regardless of how the application has decided to use processes and threads.

In reviewing why the locking implementation is the way it is, it occurred to me that all the complexity is due to not taking the lock type in the `ostree_repo_lock_pop` call. The reason it doesn't is to make the API friendly - you push the lock type you want an pop will just take it off when you say to. However, doing that requires the semi-complex stack of lock states so that you know which one to remove.

However, if you drop that requirement and make the caller specify the lock type to pop, then the tracking becomes very simple again - 2 counters and as long as the number of exclusive locks is greater than 0 than the lock is exclusive. I think that makes the whole thing much simpler and implemented it in https://github.com/dbnicholson/ostree/tree/mt-locking-alt. Let me know if that looks better to you and I can move it here.

Reply:
> Yes, in that case you could add some unlocked variant that's only available internally. That case actually violates the semantics since the transaction only has a shared lock and the expectation is that you have an exclusive lock when deleting.

Hmm.  The tombstone object is very special though - clients just test whether or not it exists.  Here a separate client with a shared lock querying the commit (that doesn't exist ) running simultaneously with another client writing that commit is racy, but in a safe way.  We're not talking about *losing* data here, conceptually it's deleting a file as part of the act of *creating* data.

My mental model here is still thinking of shared = read, exclusive = write.

Reply:
> My mental model here is still thinking of shared = read, exclusive = write.

It took me a while to not think of them that way, but I've also touched this code by far the most. Maybe we should rename the lock names before making them public? I think I only named them that way initially since those were the flock names. The semantics it's really trying to encode are to not delete things that others might be using. I think you could name those `ACCESS` and `DELETE`. Probably there are better names, though.

Reply:
There's lots of terms of art in https://en.wikipedia.org/wiki/Isolation_(database_systems)

I think shared/exclusive is better than read/write - it's just easier to think of read/write in many cases.  That said for example in Rust there's a sentiment that `&/&mut` are misnamed and it should have been `&/&excl` or so (short for "exclusive"), because `&` in some cases still allows [interior mutability](https://doc.rust-lang.org/reference/interior-mutability.html)).

Reply:
So we got into this discussion around the complexity of the state tracking, which was simplified by having the caller track which lock they acquired.  That's better, but we were also having a discussion around the *semantics* of locking that I'm not sure is resolved.  You said you wanted to change `ostree_repo_delete_object()` to acquire an exclusive lock, which makes some sense.

I do still think the right fix is either:

 - For removing *tombstone* objects to not require an exclusive lock
 - Change the pull code to call an `_unlocked()` variant

Or honestly, it's going to be an ABI hazard to change the locking semantics of any functions at all, and that may apply in this case.

Actually `_delete_object()` is also inherently very special and low level - you can easily just remove a file object referenced by a commit, etc.  So we could instead add `ostree_repo_delete_object_locked()` or so...but under what scenarios should that be used?

Were you actually planning to use this for something?

Reply:
The semantics and hazards is exactly why I started working on this again. In the case of `_delete_object()`, there are several ways to make it work. In addition to the ways you mentioned above, I also have a local branch that adds a `_delete_object_if_exists()` which simply does a `_have_object()` before calling into the real `_delete_object()` to avoid taking the lock unless you really need to. It's still racy, but for the case we're describing here it's not really different. Also just documenting that `_delete_object()` is low level and you take an exclusive lock before using it would be fine, too.

Anyways, I wasn't trying to focus on this specific case. I don't have plans to use `_delete_object()` directly. I was trying to use it as a concrete example of the general issue I'm worried about. The semantics of the repo locking without this PR are that threads operate on separate lock file descriptors. If you have multiple threads operating on the same `OstreeRepo`, then sooner or later you're likely to have a deadlocked application when someone adds an innocuous lock call in ostree's internals.

I find this to not be a theoretical issue since ostree itself becomes multithreaded in pull, one of the most common operations someone would do on an ostree repo. So, I believe this hazard is not just a corner case of someone trying to run a multithreaded ostree program, but actually pretty likely to occur. All that has to happen is someone adds a locking call somewhere in an API that pull uses. My focus on `_delete_object()` was that it's a realistic (or even desired) candidate.

With this PR the hazard reduces a lot and it actually gives the application (or even libostree itself) options. Threads will share the lock file descriptor and coordinate on locking. This is the same as if you wrote a multithreaded `flock()` program and opened the file descriptor before spawning threads. If you want the threads to act independently (e.g. a pull in one thread and a prune in another) you just open separate `OstreeRepo`s in each thread. I think this is a lot better and probably just needs some clearer documentation of the semantics.